File: examples/Pose2SLAMExample_advanced.cpp
/* ----------------------------------------------------------------------------

 * GTSAM Copyright 2010, Georgia Tech Research Corporation, 
 * Atlanta, Georgia 30332-0415
 * All Rights Reserved
 * Authors: Frank Dellaert, et al. (see THANKS for the full author list)

 * See LICENSE for the license information

 * -------------------------------------------------------------------------- */

/**
 * @file Pose2SLAMExample_advanced.cpp
 * @brief Simple Pose2SLAM Example using
 * pre-built pose2SLAM domain
 * @author Chris Beall
 */

// pull in the Pose2 SLAM domain with all typedefs and helper functions defined
#include <gtsam/slam/pose2SLAM.h>
#include <gtsam/nonlinear/LevenbergMarquardtOptimizer.h>
#include <gtsam/nonlinear/Marginals.h>
#include <gtsam/base/Vector.h>
#include <gtsam/base/Matrix.h>

#include <boost/shared_ptr.hpp>
#include <cmath>
#include <iostream>

using namespace std;
using namespace gtsam;

int main(int argc, char** argv) {
	/* 1. create graph container and add factors to it */
	pose2SLAM::Graph graph;

	/* 2.a add prior  */
	Pose2 priorMean(0.0, 0.0, 0.0); // prior at origin
	SharedDiagonal priorNoise(Vector_(3, 0.3, 0.3, 0.1)); // 30cm std on x,y, 0.1 rad on theta
	graph.addPrior(1, priorMean, priorNoise); // add directly to graph

	/* 2.b add odometry */
	SharedDiagonal odometryNoise(Vector_(3, 0.2, 0.2, 0.1)); // 20cm std on x,y, 0.1 rad on theta
	Pose2 odometry(2.0, 0.0, 0.0); // create a measurement for both factors (the same in this case)
	graph.addOdometry(1, 2, odometry, odometryNoise);
	graph.addOdometry(2, 3, odometry, odometryNoise);
	graph.print("full graph");

	/* 3. Create the data structure to hold the initial estimate to the solution
	 * initialize to noisy points */
	pose2SLAM::Values initial;
	initial.insertPose(1, Pose2(0.5, 0.0, 0.2));
	initial.insertPose(2, Pose2(2.3, 0.1, -0.2));
	initial.insertPose(3, Pose2(4.1, 0.1, 0.1));
	initial.print("initial estimate");

	/* 4.2.1 Alternatively, you can go through the process step by step
	 * Choose an ordering */
	Ordering ordering = *graph.orderingCOLAMD(initial);

	/* 4.2.2 set up solver and optimize */
	LevenbergMarquardtParams params;
	params.absoluteErrorTol = 1e-15;
	params.relativeErrorTol = 1e-15;
	params.ordering = ordering;
	LevenbergMarquardtOptimizer optimizer(graph, initial, params);

	pose2SLAM::Values result = optimizer.optimize();
	result.print("final result");

	/* Get covariances */
	Marginals marginals(graph, result, Marginals::CHOLESKY);
	Matrix covariance1 = marginals.marginalCovariance(1);
	Matrix covariance2 = marginals.marginalCovariance(2);

	print(covariance1, "Covariance1");
	print(covariance2, "Covariance2");

	return 0;
}

File: examples/matlab/Pose2SLAMExample_advanced.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% GTSAM Copyright 2010, Georgia Tech Research Corporation, 
% Atlanta, Georgia 30332-0415
% All Rights Reserved
% Authors: Frank Dellaert, et al. (see THANKS for the full author list)
% 
% See LICENSE for the license information
%
% @brief Simple robotics example using the pre-built planar SLAM domain
% @author Alex Cunningham
% @author Frank Dellaert
% @author Chris Beall
% @author Vadim Indelman
% @author Can Erdogan
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Assumptions
%  - All values are axis aligned
%  - Robot poses are facing along the X axis (horizontal, to the right in images)
%  - We have full odometry for measurements
%  - The robot is on a grid, moving 2 meters each step

%% Create keys for variables
x1 = 1; x2 = 2; x3 = 3;

%% Create graph container and add factors to it
graph = pose2SLAMGraph;

%% Add prior
% gaussian for prior
prior_model = gtsamSharedNoiseModel_Sigmas([0.3; 0.3; 0.1]);
prior_measurement = gtsamPose2(0.0, 0.0, 0.0); % prior at origin
graph.addPrior(x1, prior_measurement, prior_model); % add directly to graph

%% Add odometry
% general noisemodel for odometry
odom_model = gtsamSharedNoiseModel_Sigmas([0.2; 0.2; 0.1]);
odom_measurement = gtsamPose2(2.0, 0.0, 0.0); % create a measurement for both factors (the same in this case)
graph.addOdometry(x1, x2, odom_measurement, odom_model);
graph.addOdometry(x2, x3, odom_measurement, odom_model);

%% Add measurements
% general noisemodel for measurements
meas_model = gtsamSharedNoiseModel_Sigmas([0.1; 0.2]);

% print
graph.print('full graph');

%% Initialize to noisy points
initialEstimate = pose2SLAMValues;
initialEstimate.insertPose(x1, gtsamPose2(0.5, 0.0, 0.2));
initialEstimate.insertPose(x2, gtsamPose2(2.3, 0.1,-0.2));
initialEstimate.insertPose(x3, gtsamPose2(4.1, 0.1, 0.1));

initialEstimate.print('initial estimate');

%% set up solver, choose ordering and optimize
params = gtsamNonlinearOptimizationParameters_newDecreaseThresholds(1e-15, 1e-15);

ord = graph.orderingCOLAMD(initialEstimate);

%result = pose2SLAMOptimizer(graph,initialEstimate,ord,params);                      
%result.print('final result');

% %     
% % disp('\\\');
% % 
% % %% Optimize using Levenberg-Marquardt optimization with an ordering from colamd
result = graph.optimize(initialEstimate);
result.print('final result');

%% Get the corresponding dense matrix
ord = graph.orderingCOLAMD(result);
gfg = graph.linearize(result,ord);
denseAb = gfg.denseJacobian;

%% Get sparse matrix A and RHS b
IJS = gfg.sparseJacobian_();
Ab=sparse(IJS(1,:),IJS(2,:),IJS(3,:));
A = Ab(:,1:end-1);
b = full(Ab(:,end));
spy(A);
